examples/rl-ppo-continuous.rs

```diff
@@ -0,0 +1,191 @@
+//! Implements the reinforcement learning algorithm Proximal Policy Optimization (PPO) on random data.
+
+use dfdx::{
+    gradients::Tape,
+    nn::{
+        self,
+        modules::{Linear, ReLU, Tanh},
+        BuildModule, Module,
+    },
+    optim::{Adam, AdamConfig},
+    prelude::{smooth_l1_loss, GradientUpdate, ParamUpdater, UnusedTensors},
+    tensor::{AsArray, Cpu, HasErr, SplitTape, Tensor1D, Tensor2D},
+    tensor_ops::MeanTo,
+};
+use std::time::Instant;
+
+const BATCH_SIZE: usize = 64;
+const STATE_SIZE: usize = 4;
+const INNER_SIZE: usize = 128;
+const ACTION_SIZE: usize = 2;
+
+/// Custom model struct
+#[derive(Clone)]
+struct Network<const IN: usize, const INNER: usize, const OUT: usize> {
+    l1: (Linear<IN, INNER, f32, Cpu>, ReLU),
+    mu: (Linear<INNER, OUT, f32, Cpu>, Tanh),
+    std: (Linear<INNER, OUT, f32, Cpu>, ReLU), // TODO: should this be SoftPlus?
+    value: Linear<INNER, OUT, f32, Cpu>,
+}
```

Comment:
From what I can tell, l1 is always used as the first step and its result is the input the the three other functions. Sometimes we want the result from all functions and sometimes only one or some of them.

Is this how such a network should be modelled? I have tried looking at `07-custom-module.rs`. How on earth is the gradient and optimizer stuff supposed to fit in with this?

Reply:
Might it perhaps be as simple as `dev.sample_normal() * std + mu`?

Reply:
> Is this how such a network should be modelled? I have tried looking at `07-custom-module.rs`. How on earth is the gradient and optimizer stuff supposed to fit in with this?

This is probably the best way to model this kind of network, at least in terms of being able to use optimizers ergonomically. Your implementation of `GraidentUpdate` seems correct, as it closely mirrors the implementation of `GradientUpdate` for [transformer encoders](https://github.com/coreylowman/dfdx/blob/f9d6aa0ac02102057980e891c85eccf5c1f298dd/src/nn/transformer/encoder.rs#L120).

> Might it perhaps be as simple as `dev.sample_normal() * std + mu`?
>
> Edit: no, dist is not the supposed to be the Tensor, it is the normal distribution generator thing right? [Here](https://github.com/seungeunrho/minimalRL/blob/master/ppo-continuous.py#L138) it is used to generate the next action (a tensor).

Dfdx does not currently have an object that represents a random distribution, so you should just use `dev.sample_normal() * std + mu` in place of `dist.sample()`. This should be valid, since [`Normal`](https://pytorch.org/docs/stable/distributions.html#normal) uses `mu` as `loc` and `std` as `scale`.

Reply:
Ah, Thanks a lot.

How would I accomplish something like [this](https://github.com/seungeunrho/minimalRL/blob/master/ppo-continuous.py#L139):

```python
dist = Normal(mu, std)
a = dist.sample()
log_prob = dist.log_prob(a) # <-- This
```

and [this](https://github.com/seungeunrho/minimalRL/blob/master/ppo-continuous.py#L139)? What does [it](https://pytorch.org/docs/stable/distributions.html#torch.distributions.normal.Normal.log_prob) mean? I have read [this](https://stackoverflow.com/questions/54635355/what-does-log-prob-do) several times now, can't say I got that much wiser... :) Something about the logarithm of the probability for the value of action in the distribution? Does this mean it is not something we can do on our sample (dist), but rather have to calculate from action, `mu`, `std` and how the normal distribution behave?

Reply:
From what I understand, `log_prob` computes `log(pdf(x))`, where pdf is the [normal distribution's probability distribution function](https://en.wikipedia.org/wiki/Normal_distribution). Dfdx doesn't currently have an implementation of pdfs/cdfs of probability distributions, but it should be pretty easy to implement directly, and you can reference [pytorch's implementation](https://pytorch.org/docs/stable/_modules/torch/distributions/normal.html#Normal.log_prob).

Reply:
Again, thanks a lot :)